Add the first part of the new Light Optimization Layer JIT

kmiller68 opened the pull request:
#### ab55d8679525a3a67493f6f637ccd4e4a714afdf
<pre>
Add the first part of the new Light Optimization Layer JIT
<a href="https://bugs.webkit.org/show_bug.cgi?id=304250">https://bugs.webkit.org/show_bug.cgi?id=304250</a>
<a href="https://rdar.apple.com/166615545">rdar://166615545</a>

Reviewed by Marcus Plutowski and Yusuke Suzuki.

This patch adds the outline of the new LOL JIT. LOL differs from the
existing baseline in that we intend to do limited inter-bytecode
optimizations. Namely, we&apos;re going to try to register allocate between
bytecodes.

To make slow paths efficient (and eventually OSR exiting possible without
a huge memory regression) the plan is to make the register allocator
replayable. This works by having the register alligator be templated, so
that different backends can either generate code for the allocation or
simply replay what was allocated the first time. To make testing of the
bytecodes that I wrote possible whenever LOL JIT sees a bytecode that it
doesn&apos;t know it will just flush all registers. This obviously generates
poor code quality but as the number of bytecodes supported increases
it should matter less and less.

Each bytecode has an overload in LOL::RegisterAllocator which knows about
the number of uses, defs and scratches a given bytecode needs. The other
upside is that when tuning in the future we can precisely control the
logic in one place. e.g. if we know that an IC uses x0-x4 we can always
flush those registers and shuffle the result to x0 and any future
replayers (i.e. slow paths/OSR) would see that change in the register
state.

I also added logging so we can see the register allocation state in
disassembly, which has been very helpful in debugging.

Some of the interesting cases I had to fix for this PR are:
1) The start of the next instruction and the slow-to-hot return point
are now different. This is because the next instruction could be a jump
target (or not supported) so we have to flush everything to the stack.

2) The current Baseline JIT only calls into BytecodeIndices that actually
have a slow path. This differs from LOL JIT where we switch over every
instruction to replay the register allocator.

Canonical link: <a href="https://commits.webkit.org/304769@main">https://commits.webkit.org/304769@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a0012cf2798ce32b17f1050c5adf9f67c14184aa

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136501 "11 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8858 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47781 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144214 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/89474 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9545 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8702 "Built successfully") | [⏳ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/WPE-WK2-Tests-EWS "Waiting to run tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/89474 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139446 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/6969 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [⏳ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/API-Tests-WPE-EWS "Waiting to run tests") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/6612 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/4275 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4807 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/128458 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/115915 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-WK1-Tests-EWS "Waiting to run tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146963 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/134985 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8540 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/41078 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/112724 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8557 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/7178 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/113068 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/164/builds/6552 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/118613 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/62536 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/21044 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8588 "Built successfully") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/167765 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8307 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/72147 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/167765 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8528 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8380 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/jit/JIT.h:
```diff
@@ -187,7 +192,7 @@ namespace JSC {
         JS_EXPORT_PRIVATE static Seconds totalCompileTime();
 
-    private:
+    protected:
```
Will revert before landing.

justinmichaud commented on Source/JavaScriptCore/lol/LOLJIT.cpp:
```diff
@@ -0,0 +1,1612 @@
+        break; \
+    }
+
+    // m_replayAllocator.reset();
```
Is this expected?

kmiller68 replied:
Ah, that was left over from an earlier iteration where replay and fast allocator were the same. Will remove.

Achierius commented on Source/JavaScriptCore/lol/LOLJIT.cpp:
```diff
@@ -0,0 +1,1612 @@
+
+            for (unsigned argument = 0; argument < m_unlinkedCodeBlock->numParameters(); ++argument) {
+                // If this is a constructor, then we want to put in a dummy profiling site (to
+                // keep things consistent) but we don't actually want to record the dummy value.
```
Is it ever valid to consume this profiling site? If not, is there a way to poison it so people don't do so accidentally?

kmiller68 replied:
I'm not sure. That's code I cargo-culted over from the Baseline. I can add a FIXME though.

Achierius commented on Source/JavaScriptCore/lol/LOLJIT.cpp:
```diff
@@ -0,0 +1,1612 @@
+
+namespace JSC::LOL {
+
+namespace LOLJITInternal {
```
Nit: since this is inside JSC::LOL, could it just be called Internal, and then you refer to JSC::LOL::Internal?

kmiller68 replied:
This is more consistent with the other `XYZInternal` namespaces e.g. `namespace B3FixSSAInternal`.

Achierius commented on Source/JavaScriptCore/jit/RegisterSet.h:
```diff
@@ -328,38 +328,25 @@ class RegisterSet final {
 
-    inline constexpr iterator begin() const { return iterator(m_bits.begin()); }
-    inline constexpr iterator end() const { return iterator(m_bits.end()); }
+    inline constexpr iterator begin() const LIFETIME_BOUND { return static_cast<iterator>(m_bits.begin()); }
```
What about your patch makes this newly LIFETIME_BOUND? Or is that a pre-existing gap?

kmiller68 replied:
Just a missing annotation. Idk if it matters.

Achierius commented on Source/JavaScriptCore/lol/LOLJIT.h:
```diff
@@ -0,0 +1,335 @@
+        auto next = BytecodeIndex(m_bytecodeIndex.offset() + m_currentInstruction->size());
+        if (m_currentJumpTargetIndex < m_unlinkedCodeBlock->numberOfJumpTargets() && next.offset() == m_unlinkedCodeBlock->jumpTarget(m_currentJumpTargetIndex)) {
+            if (shouldSetFastPathResumePoint) {
+                // We need to set a resume point for slow paths to jump back to prior to flushing since the next instruction wouldn't have the flushes and we don't want to re-emit them in the slow path.
```
I think that it's worth having a comment somewhere that illustrates what the emitted code is intended to look like / what the control flow of a slow path that needs to flush would be

Achierius commented on Source/JavaScriptCore/lol/LOLRegisterAllocator.h:
```diff
@@ -0,0 +1,271 @@
+    Location locationOf(VirtualRegister operand) const { return const_cast<RegisterAllocator<Backend>*>(this)->locationOfImpl(operand); }
+    VirtualRegister bindingFor(GPRReg reg) const { return m_allocator.bindingFor(reg); }
+
+    // In general, it's somewhat important that these don't change how they allocate based on profiling data as when someone
+    // replays that profiling data could have changed and the register state they'd get would be out of sync with reality.
```
It'd be nice to have a check for this if asserts are enabled, e.g. by caching a hash of the register allocation on the basic-block and ASSERTing whenever you re-allocate thereafter

kmiller68 replied:
I added a FIXME

Achierius reviewed (approved):
Talked this over offline, answered a few of the larger questions I had: 
 - How does this work with linking cached unlinked baseline code? -> Since this won't alloc across block boundaries, it changes nothing
 - How many slow paths are there that could be impacted? -> 40-50% of bytecodes in SP3 / JS2 have a slow path, 20% (out of the remaining 50-60% that don't) are moves, and moreover calls apparently don't have slow paths
 - What blocks enabling OSRExiting? -> just correctly handling the reg-alloc replay

kmiller68 committed:
Add the first part of the new Light Optimization Layer JIT
https://bugs.webkit.org/show_bug.cgi?id=304250
rdar://166615545

Reviewed by Marcus Plutowski and Yusuke Suzuki.

This patch adds the outline of the new LOL JIT. LOL differs from the
existing baseline in that we intend to do limited inter-bytecode
optimizations. Namely, we're going to try to register allocate between
bytecodes.

To make slow paths efficient (and eventually OSR exiting possible without
a huge memory regression) the plan is to make the register allocator
replayable. This works by having the register alligator be templated, so
that different backends can either generate code for the allocation or
simply replay what was allocated the first time. To make testing of the
bytecodes that I wrote possible whenever LOL JIT sees a bytecode that it
doesn't know it will just flush all registers. This obviously generates
poor code quality but as the number of bytecodes supported increases
it should matter less and less.

Each bytecode has an overload in LOL::RegisterAllocator which knows about
the number of uses, defs and scratches a given bytecode needs. The other
upside is that when tuning in the future we can precisely control the
logic in one place. e.g. if we know that an IC uses x0-x4 we can always
flush those registers and shuffle the result to x0 and any future
replayers (i.e. slow paths/OSR) would see that change in the register
state.

I also added logging so we can see the register allocation state in
disassembly, which has been very helpful in debugging.

Some of the interesting cases I had to fix for this PR are:
1) The start of the next instruction and the slow-to-hot return point
are now different. This is because the next instruction could be a jump
target (or not supported) so we have to flush everything to the stack.

2) The current Baseline JIT only calls into BytecodeIndices that actually
have a slow path. This differs from LOL JIT where we switch over every
instruction to replay the register allocator.

Canonical link: https://commits.webkit.org/304769@main
